### DISCUSSION TITLE
Getting forward slice in headless.

### DISCUSSION THREAD

**Discussion by anzosasuke** [NationalSecurityAgency/ghidra#4133] [Q&A]
I have been trying to get the feature of Ghidra GUI of getting forward slice like "Highlight Forward Slice" from headless. I've explored DecompilerUtils().getForwardSlice(Varnode []). But I was not able to make sense of it. Can someone point me in the right direction?

Thanks in advance!!

**Comment by ghidracadabra** [comment]
The varnodes you pass that method should come from a HighFunction (which contains information produced during decompilation).  The script GraphASTScript.java has a simple example of getting a HighFunction programmatically (see the method buildAST()).  From there, one method to get at the varnodes would be to iterate over the pcode ops (via HighFunction.getPcodeOps()) and examine their inputs/outputs.  

Given a varnode from a HighFunction, out.getHigh() yields the high variable, and out.getHigh().getName() yields the name of the variable that you see in the decompiled output.  Note that the name can be null in certain cases.  Also note that you'll have to change one line the the buildAST() example to get the decompiler to do variable recovery: change
`ifc.setSimplificationStyle("normalize");` 
to 
`ifc.setSimplificationStyle("decompile");`

**Comment by anzosasuke** [comment]
Thanks for the reply @ghidracadabra . I got the output like this for `print(i.getHigh().getName())
                print(decomp.getForwardSlice(i))` 
output snippet looks like this:
local_30
[(unique, 0x100000ab, 4), (unique, 0xbf80, 4), (unique, 0xd100, 1), (stack, 0xffffffffffffffd0, 4), (register, 0x30, 8)]

Could you please tell me, in general, how do I read these varnodes?

**Reply by ghidracadabra** [reply]
The information in the tuple is (address space, offset, size).  You can find more information in Section 1 ("Introduction to P-Code") of `sleigh.html` in the language documentation included with a Ghidra distribution.  

The "unique" space is essentially a scratch space used by the decompiler and language module implementors (e.g., it's where the local variables used in a SLEIGH specification live).

Additionally, the `Varnode` class has a `toString(Language language)` method, which you can call in a script via
`vn.toString(currentProgram.getLanguage())` (where `vn` is a varnode).  This output will be easier to read.  For instance, registers will be shown using their names rather than as an offset in the "register" space.
